File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Coolstore/Services/PromoService.cs
using Ecommerce.Coolstore.Models;

namespace Ecommerce.Coolstore.Services;

public class PromoService
{
    private readonly Dictionary<string, double> _percentOffByItem = new()
    {
        ["329299"] = 0.25
    };

    public void ApplyCartItemPromotions(ShoppingCart shoppingCart)
    {
        if (shoppingCart.ShoppingCartItemList.Count == 0)
        {
            return;
        }

        foreach (var sci in shoppingCart.ShoppingCartItemList)
        {
            if (sci.Product == null)
            {
                continue;
            }

            if (_percentOffByItem.TryGetValue(sci.Product.ItemId, out var pct))
            {
                sci.PromoSavings = sci.Product.Price * pct * -1;
                sci.Price = sci.Product.Price * (1 - pct);
            }
        }
    }

    public void ApplyShippingPromotions(ShoppingCart shoppingCart)
    {
        if (shoppingCart.CartItemTotal >= 75)
        {
            shoppingCart.ShippingPromoSavings = shoppingCart.ShippingTotal * -1;
            shoppingCart.ShippingTotal = 0;
        }
    }
}

File: rgctl-tests/ecommerce-typescript/src/coolstore/services/promoService.ts
import type { ShoppingCart } from '../models/shoppingCart';

export class PromoService {
  private readonly percentOffByItem = new Map<string, number>([['329299', 0.25]]);

  applyCartItemPromotions(shoppingCart: ShoppingCart): void {
    if (!shoppingCart.shoppingCartItemList.length) {
      return;
    }
    for (const sci of shoppingCart.shoppingCartItemList) {
      if (!sci.product) {
        continue;
      }
      const pct = this.percentOffByItem.get(sci.product.itemId);
      if (pct != null) {
        sci.promoSavings = sci.product.price * pct * -1;
        sci.price = sci.product.price * (1 - pct);
      }
    }
  }

  applyShippingPromotions(shoppingCart: ShoppingCart): void {
    if (shoppingCart.cartItemTotal >= 75) {
      shoppingCart.shippingPromoSavings = shoppingCart.shippingTotal * -1;
      shoppingCart.shippingTotal = 0;
    }
  }
}
